### ISSUE TITLE
support mod train stops as well as core

### ISSUE THREAD

**Issue by slippycheeze** [leehuk/factorio-tnfp#2]
G'day.  It'd be nice to have support for mod train stops (eg: LTN, TSM, etc, custom stops)

This is simply a matter of changing the `find_entities_filtered` argument from `{name = 'train-stop'}` to `{type = 'train-stop'}`, which should work for any mod.  It certainly does in my local testing.

This needs to happen in both `stops.lua` and `gui.lua`, since they both independently use the search feature.

**Comment by slippycheeze** [comment]
...just found a second place in `stops.lua` I needed to update; apparently my checking wasn't as thorough as it should have been.  Updating that to use type rather than name fixes calling trains, though, so should be simple if you actually bother searching for all uses of it.  I didn't, and regret my life choices. :)

**Comment by leehuk** [comment]
Aye I should have some time over the weekend to run some testing in creative mode with both LTN and TSM so I can see how complex this ends up needing to be.

Specifically, I'm wondering for LTN if we dispatched to a LTN depot station whether LTN is likely to 'claim' the train and replace its schedule.  I'd expect dispatching to anything other than a depot is fine and that said it should be possible to detect depot stations via a standard circuit network check to special case them.

Similarly for TSM, I'm wondering if we dispatched a train to a 'provider' station whether TSM could end up sending it back out for a delivery.  That situation could be harder to deal with as I believe TSM relies more heavily on internal state for that kind of logic.

Gut feel is it'd probably end up working something like this:
* If its a vanilla train station, dispatch as normal.
* If its a non-depot LTN train station, dispatch as normal.
* (Potentially some custom TSM logic to dispatch as normal)
* Otherwise automatically create a temporary stop near the other mods stop and dispatch to that instead.

Effectively the presumption would be if tnfp isn't tested to be explicitly compatible, it defaults to a 'safe' (but arguably non-ideal behaviour), though we could equally do the inverse and only do the 'safe' behaviour where it knows its explicitly incompatible, or there could be a configuration option to determine which it chooses when it doesn't know for sure.

If you've any more thoughts or other mods you had in mind let me know, but overall I think it'll be achieveable as the core codes there to support whats likely needed.

**Comment by leehuk** [comment]
So for LTN I didn't realise the "is depot" signal comes into the LTN input lamp and not to the train stop itself -- so ideally I would mirror the logic directly from the LTN source code about how train stops and input lamps are connected together to be able to get to the LTN signals, but the simpler method of copying that code from the LTN source isn't permitted under the LTN codes license without Optera's approval.

So, here goes.. :)

@Yousei9 is there any possibility you could grant an exception to the LTN license, permitting the copying of any small sections of code from LTN into TNfP, that specifically relate to how to map from a known LTN train stop entity to its connected input lamp and for that code to be then released as part of this project under the MIT license please?

The intent is to ensure TNfP can safely dispatch a train to an LTN train stop entity, whilst having special handling logic so TNfP doesn't dispatch to things like LTN depot stations where the mods would start having conflicting behaviours.

I'd expect that will likely be a small code section, presumably of either special circuit logic, entity naming logic or positional/directional logic that maps the two together, but I wanted to be careful with deriving any logic at all from the LTN source before I checked with you.

**Comment by 0ptera** [comment]
I won't grant exceptions to re-release code under MIT or similar. 

LTNs lamps are connected very deeply. They are not only input for controls signals, but also read item/fluid count and act as in game status display.

To me it sounds as if all you want to know is which LTN stop is flagged as depot. 
LTN already exposes all internals through an API: https://forums.factorio.com/viewtopic.php?f=214&t=51072#p397844
Subscribe to on_stops_updated and LTN will give you all the data about stop composite entities you could ever want. Should something be missing let me know.

The following snippet provides a lookup dictionary for stops flagged as depots indexed by entity.unit_number:
```
global.ltn_depots = global.ltn_depots or {}

function OnStopsUpdated(event)
  global.ltn_depots = {} -- wipe existing depot lookup
  if event.logistic_train_stops then
    -- find all depots and add to lookup
    for stopID, ltn_stop in pairs(event.logistic_train_stops) do
      if ltn_stop.isDepot then
        global.ltn_depots[stopID] = true
      end
    end
  end
end
```

**Comment by leehuk** [comment]
I wasn't aware LTN had an event based API, so I'll look into hooking into that properly via the snippet you've provided -- much appreciated for that.

**Comment by slippycheeze** [comment]
I should have said something earlier, probaby, but I'm entirely happy with the safe default behaviour. :)

Also, when I hacked in LTN support I don't think I actually visited the depot ever, and it probably would have looked like a train being sent for LTN to manage.  Worked fine for regular stops though, including requester stops.

**Comment by leehuk** [comment]
This is all now merged and released as part of v0.4.1 which I've pushed to the mod portal.  Anything further please don't hesitate to ask!
